### PR TITLE
validator: added support for autocasting from a string

### PR DESCRIPTION
**Pull request by hrach**

### CODE REVIEW

File: src/Validator.php [hunk 1]
```diff
@@ -76,16 +81,28 @@ public function validate($data): Result
...
+						} elseif ($autoCast && ($node === '1' || $node === '0')) {
+							$node = (bool) $node;
+							$isValid = true;
+							break;
 						}
 					} elseif ($type === 'int') {
 						if (is_int($node)) {
 							$isValid = $this->validateNumber($node, $schema, $path, $errors);
 							break;
+						} elseif ($autoCast && is_string($node) && preg_match('#^-?[1-9][0-9]*+\z#', $node) === 1) {
```
Comment (JanJakes) [inline src/Validator.php]: This wouldn't accept `0`, right? There is also `filter_var`... maybe?

File: src/Validator.php [hunk 2]
```diff
@@ -76,16 +81,28 @@ public function validate($data): Result
...
+							$node = (int) $node;
+							$isValid = $this->validateNumber($node, $schema, $path, $errors);
+							break;
 						}
 					} elseif ($type === 'float') {
 						if (is_float($node)) {
 							$isValid = $this->validateNumber($node, $schema, $path, $errors);
 							break;
+						} elseif ($autoCast && is_string($node) && preg_match('#^-?[1-9][0-9]*+(\.[0-9]+)?\z#', $node) === 1) {
```
Comment (JanJakes) [inline src/Validator.php]: This also wouldn't accept `0` and there are also other float notations (1e12, etc.), but I'm not sure if we want to support those. Here `is_numeric()` would probably do the job.

File: src/Validator.php [hunk 3]
```diff
@@ -29,11 +29,16 @@ public function __construct(array $schema, bool $failFast = false, callable $ref
...
 
 
-	public function validate($data): Result
+	public function validate($data, bool $autoCast = false): Result
 	{
+		$outData = $data;
 		$errors = [];
-		$stack = [[$this->schema, $data, '/']];
-		while (list ($schema, $node, $path) = array_pop($stack)) {
+		$stack = [[$this->schema, & $outData, '/']];
```
Comment (JanJakes) [inline src/Validator.php:36]: Why so many `&`? It's kind of hard to read.

File: src/Validator.php [hunk 4]
```diff
@@ -29,11 +29,16 @@ public function __construct(array $schema, bool $failFast = false, callable $ref
 	}
 
 
-	public function validate($data): Result
+	public function validate($data, bool $autoCast = false): Result
```
Comment (JanTvrdik) [inline src/Validator.php]: btw: this is often called coercion, e.g. https://github.com/justinrainbow/json-schema#type-coercion

**Comment by hrach** [comment]
Guys, thanks for the review.

**Pull request merged**